src/omp/exec_dist.f90

```diff
@@ -61,5 +61,107 @@ subroutine exec_dist_tds_compact( &
 
    end subroutine exec_dist_tds_compact
 
+
+   subroutine exec_dist_transeq_compact(&
+      rhs, du, dud, d2u, &
+      du_send_s, du_send_e, du_recv_s, du_recv_e, &
+      dud_send_s, dud_send_e, dud_recv_s, dud_recv_e, &
+      d2u_send_s, d2u_send_e, d2u_recv_s, d2u_recv_e, &
+      u, u_recv_s, u_recv_e, &
+      ud, ud_recv_s, ud_recv_e, &
+      tdsops_du, tdsops_dud, tdsops_d2u, nu, nproc, pprev, pnext, n_block)
+
+      implicit none
+
+      ! du = d(u)
+      real(dp), dimension(:, :, :), intent(out) :: rhs, du, dud, d2u
+
+      ! The ones below are intent(out) just so that we can write data in them,
+      ! not because we actually need the data they store later where this
+      ! subroutine is called. We absolutely don't care about the data they pass back
+      real(dp), dimension(:, :, :), intent(out) :: &
+         du_send_s, du_send_e, du_recv_s, du_recv_e
+      real(dp), dimension(:, :, :), intent(out) :: &
+         dud_send_s, dud_send_e, dud_recv_s, dud_recv_e
+      real(dp), dimension(:, :, :), intent(out) :: &
+         d2u_send_s, d2u_send_e, d2u_recv_s, d2u_recv_e
+
+      real(dp), dimension(:, :, :), intent(in) :: u, u_recv_s, u_recv_e
+      real(dp), dimension(:, :, :), intent(in) :: ud, ud_recv_s, ud_recv_e
+
+      type(tdsops_t), intent(in) :: tdsops_du, tdsops_dud, tdsops_d2u
+      real(dp) :: nu
+      integer, intent(in) :: nproc, pprev, pnext
+      integer, intent(in) :: n_block
+
+      integer :: n_data
+      integer :: k, i, j
+
+      n_data = SZ*n_block
+
+      !$omp parallel do
+      do k = 1, n_block
+         call der_univ_dist( &
+            du(:, :, k), du_send_s(:, :, k), du_send_e(:, :, k), u(:, :, k), &
+            u_recv_s(:, :, k), u_recv_e(:, :, k), &
+            tdsops_du%coeffs_s, tdsops_du%coeffs_e, tdsops_du%coeffs, tdsops_du%n, &
+            tdsops_du%dist_fw, tdsops_du%dist_bw, tdsops_du%dist_af &
+            )
+
+         call der_univ_dist( &
+            d2u(:, :, k), d2u_send_s(:, :, k), d2u_send_e(:, :, k), u(:, :, k), &
+            u_recv_s(:, :, k), u_recv_e(:, :, k), &
+            tdsops_d2u%coeffs_s, tdsops_d2u%coeffs_e, tdsops_d2u%coeffs, tdsops_d2u%n, &
+            tdsops_d2u%dist_fw, tdsops_d2u%dist_bw, tdsops_d2u%dist_af &
+            )
+
+         call der_univ_dist( &
+            dud(:, :, k), dud_send_s(:, :, k), dud_send_e(:, :, k), ud(:, :, k), &
```

Comment:
Because we pass a very small section of the domain into the der_univ_dist, the multiplication between convective velocity and primary velocity fields can be carried out just here only for the bit that we're passing. Just to give an idea, its only around 64 KiB when n=1024 and SZ=8.
So just allocate a 2D array with size (SZ, n), then just before calling der_univ_dist third time do
`temp = convective*u`, where convective is either `u`, `v` or `w`. Then the `temp` can be passed to der_univ_dist instead of `ud`. It might worth to do the multiplication with an explicit loop and simd directives.

Reply:
ah yes, ok. I always assume we are playing with large arrays. That will simplify things